api/catalog/api/migrations/0052_relational_fields.py

```diff
@@ -0,0 +1,157 @@
+# Generated by Django 4.1.2 on 2022-12-12 05:54
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+REPORTING_TABLES = [
+    "api_deletedaudio",
+    "api_deletedimage",
+    "api_matureaudio",
+    "api_matureimage",
+    "nsfw_reports",
+    "nsfw_reports_audio",
+]
+forward_query = "".join([
+    f"UPDATE {table} SET media_obj_id = identifier;"
+    for table in REPORTING_TABLES
+])
+reverse_query = "".join([
+    f"UPDATE {table} SET identifier = media_obj_id;"
+    for table in REPORTING_TABLES
+])
```

Comment:
@dhruvkb Is there a way to avoid these data migration queries? Because they're on tables that are pretty small I don't think it will be a big issue, but in practice we'd like to avoid these whenever possible.

I've read through the migration a few times and I do not really understand what is going on in it. I'm confused by the `identifier` field getting altered and then removed from the report tables. The changes to `media_obj` are confusing as well...

When I run `makemigrations` locally, after deleting 0052, I get prompted to enter a default value for `media_obj` because it cannot be null (I guess because it is configured to be a PK?)...

I like the changes generally being made to the tables, but I'm a little worried about not understanding this migration and the differences when I try to generate a migration locally :confused:

Reply:
Update: when I give a blank default (which I don't know if it will work due to the field type), it generates the following migration:

```py
# Generated by Django 4.1.3 on 2022-12-19 22:42

from django.db import migrations, models
import django.db.models.deletion


class Migration(migrations.Migration):

    dependencies = [
        ('api', '0051_delete_sourcelogo'),
    ]

    operations = [
        migrations.RemoveField(
            model_name='audioreport',
            name='identifier',
        ),
        migrations.RemoveField(
            model_name='deletedaudio',
            name='identifier',
        ),
        migrations.RemoveField(
            model_name='deletedimage',
            name='identifier',
        ),
        migrations.RemoveField(
            model_name='imagereport',
            name='identifier',
        ),
        migrations.RemoveField(
            model_name='matureaudio',
            name='identifier',
        ),
        migrations.RemoveField(
            model_name='matureimage',
            name='identifier',
        ),
        migrations.AddField(
            model_name='audio',
            name='audioset',
            field=models.ForeignObject(from_fields=['audio_set_foreign_identifier', 'provider'], null=True, on_delete=django.db.models.deletion.DO_NOTHING, to='api.audioset', to_fields=['foreign_identifier', 'provider']),
        ),
        migrations.AddField(
            model_name='audioreport',
            name='media_obj',
            field=models.ForeignKey(db_constraint=False, default='', help_text='The reference to the audio being reported.', on_delete=django.db.models.deletion.DO_NOTHING, related_name='audio_report', to='api.audio', to_field='identifier'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='deletedaudio',
            name='media_obj',
            field=models.OneToOneField(db_constraint=False, default='', help_text='The reference to the deleted audio.', on_delete=django.db.models.deletion.DO_NOTHING, primary_key=True, related_name='deleted_audio', serialize=False, to='api.audio', to_field='identifier'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='deletedimage',
            name='media_obj',
            field=models.OneToOneField(db_constraint=False, default='', help_text='The reference to the deleted image.', on_delete=django.db.models.deletion.DO_NOTHING, primary_key=True, related_name='deleted_image', serialize=False, to='api.image', to_field='identifier'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='imagereport',
            name='media_obj',
            field=models.ForeignKey(db_constraint=False, default='', help_text='The reference to the image being reported.', on_delete=django.db.models.deletion.DO_NOTHING, related_name='image_report', to='api.image', to_field='identifier'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='matureaudio',
            name='media_obj',
            field=models.OneToOneField(db_constraint=False, default='', help_text='The reference to the mature audio.', on_delete=django.db.models.deletion.DO_NOTHING, primary_key=True, related_name='mature_audio', serialize=False, to='api.audio', to_field='identifier'),
            preserve_default=False,
        ),
        migrations.AddField(
            model_name='matureimage',
            name='media_obj',
            field=models.OneToOneField(db_constraint=False, default='', help_text='The reference to the mature image.', on_delete=django.db.models.deletion.DO_NOTHING, primary_key=True, related_name='mature_image', serialize=False, to='api.image', to_field='identifier'),
            preserve_default=False,
        ),
    ]
```

I _think_ the difference between this version and the one in the PR currently is that the one you've generated moves the data over from identifier over to media_obj? Is that correct? The _way_ the migration does that is still confusing to me (I don't know what the AlterFields on `identifier` are for) though.

And I kind of wonder if we can avoid this complex migration to the DB queries specific work and do improvements to the admin as a separate issue and PR altogether? Do you think it is possible to split them out or are they both needed at the same time? It would make reviewing this easier if the admin improvements (which are very welcome!) were not included, especially if they're the source of the migrations :thinking:

Reply:
@sarayourfriend the changes were not for the admin but for making FK relations from the report, mature and deleted models to the media models. If we can work with calling the reference field as `identifier`, the migration gets a lot simpler as in 523338d.

Reply:
I also verified with the `sqlmigrate` command that the migration 0052 is largely a no-op! These are the only SQL changes in the migration:

```sql
BEGIN;
--
-- Add field audioset to audio
-- (no-op)
--
-- Alter field identifier on audioreport
CREATE INDEX "nsfw_reports_audio_identifier_ebe3a079" ON "nsfw_reports_audio" ("identifier");
--
-- Alter field identifier on deletedaudio
-- (no-op)
-- Alter field identifier on deletedimage
-- (no-op)
--
-- Alter field identifier on imagereport
CREATE INDEX "nsfw_reports_identifier_f0374e03" ON "nsfw_reports" ("identifier");
--
-- Alter field identifier on matureaudio
-- (no-op)
-- Alter field identifier on matureimage
-- (no-op)
COMMIT;
```